util/topsql/reporter/subscriber.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// TopSQLPublisher implements TopSQLPublisher.
+//
+// If a client subscribes to TopSQL records, the TopSQLPublisher is responsible for registering them to the reporter.
+// Then the reporter sends data to the client periodically.
+type TopSQLPublisher struct {
+	decodePlan     planBinaryDecodeFunc
+	clientRegistry *ReportClientRegistry
+}
+
+// NewTopSQLPublisher creates a new TopSQLPublisher.
+func NewTopSQLPublisher(
+	decodePlan planBinaryDecodeFunc,
+	clientRegistry *ReportClientRegistry,
+) *TopSQLPublisher {
+	return &TopSQLPublisher{
+		decodePlan:     decodePlan,
+		clientRegistry: clientRegistry,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPublisher{}
+
+// Subscribe registers clients to the reporter and redirects data received from reporter
+// to subscribers associated with those clients.
+func (t *TopSQLPublisher) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	sc := newSubClient(stream, t.decodePlan)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go sc.run(&wg)
+
+	t.clientRegistry.register(sc)
+
+	wg.Wait()
+	return nil
+}
+
+type subClient struct {
+	stream   tipb.TopSQLPubSub_SubscribeServer
+	sendTask chan struct {
+		data    reportData
+		timeout time.Duration
+	}
+	isDown *atomic.Bool
+
+	decodePlan planBinaryDecodeFunc
+}
+
+func newSubClient(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+	decodePlan planBinaryDecodeFunc,
+) *subClient {
+	sendTask := make(chan struct {
+		data    reportData
+		timeout time.Duration
+	})
+	return &subClient{
+		stream:   stream,
+		sendTask: sendTask,
+		isDown:   atomic.NewBool(false),
+
+		decodePlan: decodePlan,
+	}
+}
+
+func (s *subClient) run(wg *sync.WaitGroup) {
+	defer func() {
+		wg.Done()
+		s.isDown.Store(true)
+	}()
+
+	for task := range s.sendTask {
+		ctx, cancel := context.WithTimeout(context.Background(), task.timeout)
+		start := time.Now()
+
+		var err error
+		doneCh := make(chan struct{})
+		go func(
```

Comment:
use `go util.WithRecovery`

Reply:
`go util.WithRecovery...`

Reply:
Use `util.WithRecovery` already. The `func(task sendTask)` is to capture the `task`

Reply:
em... `go util.WithRecovery` is the default usage case....